### ISSUE TITLE
Bug: (Parsing error on unknown file extension)

### ISSUE THREAD

**Issue by TkinterinShanghai** [typescript-eslint/typescript-eslint#4763]
- [&check;] I have tried restarting my IDE and the issue persists.
- [&check;] I have updated to the latest version of the packages.
- [&check;] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

Install the Quokka and ESLint extensions in VSCode, initialize a nodejs project, run `npm install --save-dev eslint @typescript-eslint/parser @typescript-eslint/eslint-plugin` add the `eslintrc.json` below to the current working directory, go to the VSCode command palette and use the `Quokka.js: New File ` command. The error further down below will show

```JSON
{
  "env": {
    "node": true,
    "es2021": true
  },
  "extends": ["eslint:recommended", "plugin:@typescript-eslint/recommended"],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "project": "./tsconfig.json",
    "ecmaVersion": "latest",
    "sourceType": "module"
  },
  "plugins": ["@typescript-eslint"],
}
```

**Expected Result**

Not ESLint errors

**Actual Result**

```
Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: test[object Object].
The extension for the file () is non-standard. You should add "parserOptions.extraFileExtensions" to your config.eslint
```

Running eslint on the project doesnt throw an exception, because the file is not part of the project I suppose

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `5.16.0` |
| `TypeScript`                | `4.6.3` |
| `ESLint`                    | `8.12.0` |
| `node`                      | `16.12.0` |

**Comment by bradzacher** [comment]
I assume this is just due to some weirdness in the way that the quokka extension "creates" files that interacts weirdly with how the eslint extension lints files.

We just parse what we're given - we have no control over how the consumer (in this case the eslint extension) calls us. If we're given a bad path then there's not a lot we can do about it.

**Comment by TkinterinShanghai** [comment]
I got a response on another Issue regarding how Quokka creates these files. The issue can be found [here](https://github.com/wallabyjs/quokka/issues/764). Maybe this could be fixed somehow with that information

**Comment by smcenlly** [comment]
@bradzacher - could you please re-open this issue? I believe it's a problem that needs to be fixed. I'm not entirely sure whether it's a `typescript-eslint` issue or something upstream.

I'm from the team that creates Quokka. Quokka doesn't do anything special to create files in VS Code, it uses the VS Code API. The VS Code API code below causes the same problem (outside of Quokka).

```javascript
  vscode.workspace
    .openTextDocument({ language: 'typescript', content: '' })
    .then(document => vscode.window.showTextDocument(document));
```

In fact, you can get the same issue without any special extension by running the VS Code command `File New Untitled File` and then running the `Change Language Mode` command and selecting `TypeScript`, and then start writing code.

**Comment by bradzacher** [comment]
We are given zero context by ESLint. We just get the file path and the file contents. That's it.

We trust that the file path is "correct" and references a file on disk. We have to because we don't know otherwise.

We can't do things like FS checks because virtual file systems are a thing and can be setup outside our tooling. We just trust that the input we are given is correct.

So if we are told to parse a file and the path is not real - all we know is that the path does not exist within any of the tsconfigs the user provides.
This means we only really have one option - bail out because we cannot get types for the file.

If we don't bail out then downstream lint rules will execute, trusting that they were given type information. This will cause the lint run to act with undefined behaviour. There could be crashes. There could be false positives on every token. There could be nothing. We don't know what will happen.

We do not want to enter undefined behaviour. So bailing out is the best option.

**Comment by smcenlly** [comment]
Thanks @bradzacher - it makes sense. I'm guessing that there's a problem somewhere in https://github.com/Microsoft/vscode-eslint or https://github.com/Microsoft/vscode-languageserver-node when it comes to handling new files and passing them to eslint.

**Comment by bradzacher** [comment]
Not really a bug - a limitation of ours.

In general - most of the eslint ecosystem doesn't have this limitation - all it needs is the code itself! Which means it's completely fine for the extension to call eslint with the "not yet real" file.

However as mentioned - for type-aware linting we need a file on disk for it to work. That requirement is mostly unique to our parser with type info turned on.

I don't think there's really a solution here. Even if we can uniquely identify the "fake" file, what can we do aside from throw..?

In our tests we work around this by having a file physically on disk for us to reference.

Quick spitball for a solution would be to somehow make a fake file which TS recognises that we can then use as our anchor. However I don't know if this is even possible. If it is it's likely going to be A LOT of effort - more effort than is worth investing  for this very, very rare case that can be worked around by just saving the file.
